### PR TITLE
Auth0 protected routes

### PR DESCRIPTION
**Pull request by noah-milliken**
CDW-30 Feat added Authentication guard to Routes with login and logout buttons redirecting user to Auth0 login/sign up page.

### CODE REVIEW

**Review by Cililocwee — DISMISSED** [review]
Hey Noah, did you mean to take out the Router wrapper in App.tsx? And make sure you merge into your branch and take care of the conflicts!

File: frontend/src/App.tsx [hunk 1]
```diff
@@ -1,13 +1,18 @@
 import { useEffect, useState } from "react";
-import { getAllUsers } from "./requests/getAllUsers";
-import { userData } from "./requests/types";
+import { getAllUsers } from "./theme/requests/getAllUsers";
+import { userData } from "./theme/requests/types";
```
Comment (jtabba) [inline frontend/src/App.tsx]: Why were these moved?

File: frontend/src/App.tsx [hunk 2]
```diff
@@ -26,17 +31,30 @@ const App = () => {
 	});
 	// ***PURELY FOR EXAMPLE***
 
+	const { isLoading } = useAuth0()
+
+	if(isLoading){
+		return(
+			<div>
+				<Spinner />
+			</div>
```
Comment (jtabba) [inline frontend/src/App.tsx]: Use Chakra for UI (i.e. `Box` would be suitable here). Read the Charka docs if you're unfamiliar with their components

File: frontend/src/App.tsx [hunk 3]
```diff
@@ -26,17 +31,30 @@ const App = () => {
...
 					<Route path="home" element={<Home />} />
 					<Route path="explore" element={<h1>Explore</h1>} />
 					<Route path="products" element={<h1>Products</h1>} />
 					<Route path="mentors" element={<h1>Mentors</h1>} />
+					<Route path="profile" element={<AuthenticationGuard componentAsProps={Profile} /> } />
 				</Route>
+
 			</Routes>
-		</Router>
+
```
Comment (jtabba) [inline frontend/src/App.tsx]: Why was the router removed?

File: frontend/src/components/authentication/Auth0ProviderWithNavigate.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import { AppState, User, Auth0Provider } from "@auth0/auth0-react";
+import { FC } from "react";
+import { useNavigate } from "react-router-dom";
+// Make a component that accepts children. 
+
+export interface WithChildren<T = React.ReactNode> {
```
Comment (jtabba) [inline frontend/src/components/authentication/Auth0ProviderWithNavigate.tsx]: Move this to `types.ts`

File: frontend/src/components/authentication/Auth0ProviderWithNavigate.tsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    children?: T;
+}
+
+export const Auth0ProviderWithNavigate: FC<WithChildren> = ({ children }) => {
+    const navigate = useNavigate()
+
+    
+    const DOMAIN = process.env.REACT_APP_AUTH0_DOMAIN!;
+    const CLIENT_ID = process.env.REACT_APP_AUTH0_CLIENT_ID!;
+    const REDIRECT_URI = process.env.REACT_APP_AUTH0_CALLBACK_URL!;
```
Comment (jtabba) [inline frontend/src/components/authentication/Auth0ProviderWithNavigate.tsx]: We can declare these outside of the component scope for better organisation

File: frontend/src/components/authentication/Auth0ProviderWithNavigate.tsx [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        appState?: AppState,
+        user?: User
+    ) => {
+        navigate(appState?.returnTo || window.location.pathname);
+    };
+    
+   
+    if(!(DOMAIN && CLIENT_ID && REDIRECT_URI)){
+    return null
+}
```
Comment (jtabba) [inline frontend/src/components/authentication/Auth0ProviderWithNavigate.tsx:24]: Install Prettier so your code is automatically formatted on save

File: frontend/src/components/authentication/Auth0ProviderWithNavigate.tsx [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+export const Auth0ProviderWithNavigate: FC<WithChildren> = ({ children }) => {
+    const navigate = useNavigate()
+
+    
+    const DOMAIN = process.env.REACT_APP_AUTH0_DOMAIN!;
+    const CLIENT_ID = process.env.REACT_APP_AUTH0_CLIENT_ID!;
+    const REDIRECT_URI = process.env.REACT_APP_AUTH0_CALLBACK_URL!;
+
+    const onRedirectCallback: (appState?: AppState | undefined, user?: User | undefined) => void = (
```
Comment (jtabba) [inline frontend/src/components/authentication/Auth0ProviderWithNavigate.tsx]: Is the void return required here?

File: frontend/src/components/authentication/SignInOrOut.tsx [hunk 8]
```diff
@@ -5,17 +5,38 @@ export const AuthButtons = () => {
 	const { loginWithRedirect, isAuthenticated, logout } = useAuth0();
 
 	const handleLogin = () => {
-		loginWithRedirect();
+		loginWithRedirect({
+			appState: {
+				returnTo: '/profile',
```
Comment (jtabba) [inline frontend/src/components/authentication/SignInOrOut.tsx]: Let's redirect to `/home` on sign in

File: frontend/src/components/authentication/authenticationGuard.tsx [hunk 9]
```diff
@@ -0,0 +1,21 @@
+import React, { ComponentType } from "react";
+import { withAuthenticationRequired } from "@auth0/auth0-react";
+import { Spinner } from '@chakra-ui/react'
+
+interface ComponentProps {
+    componentAsProps: ComponentType<object>
+}
```
Comment (jtabba) [inline frontend/src/components/authentication/authenticationGuard.tsx:8]: Give the interface a more specific name (i.e. `IAuthenticationGuard` or `AuthenticationGuardProps`)

File: frontend/src/components/authentication/authenticationGuard.tsx [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+
+export const AuthenticationGuard: React.FC<ComponentProps> = ({componentAsProps}) => {
+    const Component = withAuthenticationRequired(componentAsProps, {
+        onRedirecting: ()=> (
+            <div className="page-layout">
+               <Spinner />
+            </div>
```
Comment (jtabba) [inline frontend/src/components/authentication/authenticationGuard.tsx]: Use Chakra

File: frontend/src/components/header/NavBar.tsx [hunk 11]
```diff
@@ -1,9 +1,15 @@
...
-export const NavBar = () => (
+
+export const NavBar = () => {
+
+	return (
 	<>
 		<Link to={"/explore"}>Explore</Link>
 		<Link to={"/products"}>Product</Link>
 		<Link to={"/mentors"}>Mentors</Link>
+		<Link to={"/profile"}>Profile</Link>
```
Comment (jtabba) [inline frontend/src/components/header/NavBar.tsx]: We'll link `/profile` inside the drop down menu instead. Remove for now

**Review by jtabba — CHANGES_REQUESTED** [review]
Great implementation overall Noah. Left some comments and am curious to hear why things that are not theme related are being move into `/theme`?

File: frontend/src/components/authentication/authenticationGuard.tsx (hunk 10 above)
Comment (noah-milliken) [reply frontend/src/components/authentication/authenticationGuard.tsx]: I used the Chakra Spinner here. Is there a chakra div element as well?

File: frontend/src/components/authentication/Auth0ProviderWithNavigate.tsx (hunk 6 above)
Comment (noah-milliken) [reply frontend/src/components/authentication/Auth0ProviderWithNavigate.tsx:24]: OK, I have prettier. But when I run it, it alters all of the files in the Repo.

**Comment by noah-milliken** [comment]
Prettier updated some formatting, I hope this doesn't Cause any conflicts.

**Comment by jtabba** [comment]
You need to merge prod into your feature branch before I can approve

**Comment by noah-milliken** [comment]
That should do it.

**Review by jtabba — APPROVED** [review]

**Pull request merged**